src/Model/Item.php

```diff
@@ -1497,88 +1480,50 @@ private static function getDuplicateID(array $item)
 	}
 
 	/**
-	 * Fetch parent data for the given item array
+	 * Fetch top-level parent data for the given item array
 	 *
 	 * @param array $item
 	 * @return array item array with parent data
+	 * @throws \Exception
 	 */
-	private static function getParentData(array $item)
+	private static function getTopLevelParent(array $item)
 	{
-		// find the parent and snarf the item id and ACLs
-		// and anything else we need to inherit
-
-		$fields = ['uri', 'parent-uri', 'id', 'deleted',
+		$fields = ['uid', 'uri', 'parent-uri', 'id', 'deleted',
 			'allow_cid', 'allow_gid', 'deny_cid', 'deny_gid',
 			'wall', 'private', 'forum_mode', 'origin', 'author-id'];
-		$condition = ['uri' => $item['parent-uri'], 'uid' => $item['uid']];
+		$condition = ['uri' => $item['thr-parent'], 'uid' => $item['uid']];
 		$params = ['order' => ['id' => false]];
 		$parent = self::selectFirst($fields, $condition, $params);
 
 		if (!DBA::isResult($parent)) {
-			Logger::info('item parent was not found - ignoring item', ['parent-uri' => $item['parent-uri'], 'uid' => $item['uid']]);
+			Logger::notice('item parent was not found - ignoring item', ['thr-parent' => $item['thr-parent'], 'uid' => $item['uid']]);
 			return [];
-		} else {
-			// is the new message multi-level threaded?
-			// even though we don't support it now, preserve the info
-			// and re-attach to the conversation parent.
-			if ($parent['uri'] != $parent['parent-uri']) {
-				$item['parent-uri'] = $parent['parent-uri'];
-
-				$condition = ['uri' => $item['parent-uri'],
-					'parent-uri' => $item['parent-uri'],
-					'uid' => $item['uid']];
-				$params = ['order' => ['id' => false]];
-				$toplevel_parent = self::selectFirst($fields, $condition, $params);
-
-				if (DBA::isResult($toplevel_parent)) {
-					$parent = $toplevel_parent;
-				}
-			}
-
-			$item['parent']        = $parent['id'];
-			$item["deleted"]       = $parent['deleted'];
-			$item["allow_cid"]     = $parent['allow_cid'];
-			$item['allow_gid']     = $parent['allow_gid'];
-			$item['deny_cid']      = $parent['deny_cid'];
-			$item['deny_gid']      = $parent['deny_gid'];
-			$item['parent_origin'] = $parent['origin'];
+		}
 
-			// Don't federate received participation messages
-			if ($item['verb'] != Activity::FOLLOW) {
-				$item['wall'] = $parent['wall'];
-			} else {
-				$item['wall'] = false;
-			}
+		if ($parent['uri'] == $parent['parent-uri']) {
+			return $parent;
+		}
 
-			/*
-			 * If the parent is private, force privacy for the entire conversation
-			 * This differs from the above settings as it subtly allows comments from
-			 * email correspondents to be private even if the overall thread is not.
-			 */
-			if ($parent['private']) {
-				$item['private'] = $parent['private'];
-			}
+		$condition = ['uri' => $parent['parent-uri'],
+			'parent-uri' => $parent['parent-uri'],
+			'uid' => $parent['uid']];
+		// We select wall = 1 in priority for top level permission checks
+		$params = ['order' => ['wall' => true]];
```

Comment:
This doesn't look right to me. We do have a parent item for this user and additionally we **might** have a parent where `wall` is true. These will very often be two different item arrays.

Reply:
I don't understand what I should change then.

Reply:
I now created a separate query to retrieve the local parent top level item if any.